gk/fib.c

```diff
@@ -1664,3 +1666,329 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+
+	eth_tbl = rte_calloc(NULL, neigh_ht->tbl_size,
+		sizeof(struct gk_fib_ether_dump_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	index = rte_hash_iterate(neigh_ht->hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		struct ether_cache *eth_cache = &neigh_ht->cache_tbl[index];
+		entry = &eth_tbl[num_entries++];
+
+		fillup_gk_fib_ether_dump_entry(eth_cache, entry);
+
+		index = rte_hash_iterate(neigh_ht->hash_table,
+			(void *)&key, &data, &next);
+	}
+
+	dentry->num_ether_entries = num_entries;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gk_fib_dump_entry(struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret = -1;
+	dentry->action = fib->action;
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		ret = fillup_grantor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		ret = fillup_gateway_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		ret = fillup_neighbor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_DROP:
+		ret = 0;
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+static int
+list_ipv4_fib_entries(lua_State *l, struct gk_lpm *ltbl,
+	char *handler, void *arg)
+{
+	int index;
+	struct gk_fib *fib;
+	const struct rte_lpm_rule *re4;
+	struct rte_lpm_iterator_state state;
+	struct gk_fib_dump_entry dentry;
+	int ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize the lpm rule iterator state at %s!\n",
+			__func__);
+		goto out;
+	}
+
+	index = rte_lpm_rule_iterate(&state, &re4);
+	while (index >= 0) {
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv4;
+		dentry.addr.ip.v4.s_addr = htonl(re4->ip);
+		dentry.len = state.depth;
+		fib = &ltbl->fib_tbl[re4->next_hop];
+		ret = fillup_gk_fib_dump_entry(&dentry, fib);
+		if (ret < 0)
+			goto out;
+
+		lua_getglobal(l, handler);
+		lua_pushlightuserdata(l, &dentry);
+		if (arg == NULL)
+			lua_pushnil(l);
+		else
+			lua_pushlightuserdata(l, arg);
+
+		if (lua_pcall(l, 2, 1, 0) != 0) {
+			lua_error(l);
+			rte_free(dentry.ether_tbl);
+			ret = -1;
+			goto out;
+		}
+
+		rte_free(dentry.ether_tbl);
+
+		index = rte_lpm_rule_iterate(&state, &re4);
+	}
+
+out:
+	return ret;
+}
+
+static int
+list_ipv6_fib_entries(lua_State *l, struct gk_lpm *ltbl,
+	char *handler, void *arg)
+{
+	int index;
+	struct gk_fib *fib;
+	struct rte_lpm6_rule re6;
+	struct rte_lpm6_iterator_state state6;
+	struct gk_fib_dump_entry dentry;
+	int ret = rte_lpm6_iterator_state_init(ltbl->lpm6, NULL, 0, &state6);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize the lpm6 rule iterator state at %s!\n",
+			__func__);
+		goto out;
```

Comment:
We are in a Lua C function here, so instead of logging, just raise a Lua error. This comment applies to a number of places in the code, so please patiently review the code to raise Lua errors.

Notice that the Lua errors are going to be much more meaningful since they will be returned as a reply to the script being run instead of going to the general log that reports everything about Gatekeeper/Grantor.